tools/setup_helpers/cudnn.py

```diff
@@ -16,26 +20,38 @@ def find_cudnn_version(cudnn_lib_dir):
     candidate_names = list(glob.glob(os.path.join(cudnn_lib_dir, 'libcudnn*')))
     candidate_names = [os.path.basename(c) for c in candidate_names]
 
-    # suppose version is MAJOR.MINOR.PATCH, all numbers
-    version_regex = re.compile('\d+\.\d+\.\d+')
-    candidates = [c.group() for c in map(version_regex.search, candidate_names) if c]
-    if len(candidates) > 0:
-        # normally only one will be retrieved, take the first result
-        return candidates[0]
-
-    # if no candidates were found, try MAJOR.MINOR
-    version_regex = re.compile('\d+\.\d+')
-    candidates = [c.group() for c in map(version_regex.search, candidate_names) if c]
-    if len(candidates) > 0:
-        return candidates[0]
-
-    # if no candidates were found, try MAJOR
-    version_regex = re.compile('\d+')
-    candidates = [c.group() for c in map(version_regex.search, candidate_names) if c]
-    if len(candidates) > 0:
-        return candidates[0]
-
-    return 'unknown'
+    # suppose lib file is is libcudnn.so.MAJOR.MINOR.PATCH, all numbers
+    version_regex = re.compile('.so.(\d+\.\d+\.\d+)$')
+    candidates = [c.group(1) for c in map(version_regex.search, candidate_names) if c]
+
+    # libcudnn.so.MAJOR.MINOR
+    version_regex = re.compile('.so.(\d+\.\d+)$')
+    candidates += [c.group(1) for c in map(version_regex.search, candidate_names) if c]
+
+    # libcudnn.so.MAJOR
+    version_regex = re.compile('.so.(\d+)$')
+    candidates += [c.group(1) for c in map(version_regex.search, candidate_names) if c]
+
+    if len(candidates) == 0:
+        return 'unknown'
+
+    # Each candidate represented as list, eg 6.0.21 -> [6, 0, 21]
+    candidates = [[int(x) for x in c.split('.')] for c in candidates]
+
+    # From candidates, take the most recent, then most detailed version string
```

Comment:
just curious: can we not do something like:
```
candidates = ['libcudnn.so.6.0.21', 'libcudnn.so.5', 'libcudnn.so.5.1.21', 'libcudnn.so', 'libcudnn.so.6']
version = sorted(candidates, reverse=True)[0]
```

Reply:
I verified locally on my python interpreter that this should work too

Reply:
also, I remember that @soumith was mentioning that we need to error out if someone has cudnn<6 , perhaps we can add a one line check for that as well?

Reply:
No, this will miscompare if you have `libcudnn.so.10` and `libcudnn.so.5`. Also, I'm pretty sure we already have the cudnn<6 check somewhere else.

Reply:
I think Lu had cudnn 5 and was building fine with it and then ran into issues with CUDNN_DETERMINISM etc. So it seems like the check might be missing in few places. 

we don't have libcudnn.so.10 lol we are still at v7